server/data_cxg/cxg_adaptor.py

```diff
@@ -200,15 +204,44 @@ def get_colors(self):
         meta = self.open_array("cxg_group_metadata").meta
         return json.loads(meta["cxg_category_colors"]) if "cxg_category_colors" in meta else dict()
 
+    def __remap_indices(self, coord_range, coord_mask, coord_data):
```

Comment:
to my eye, this is challenging to understand/validate, and uses a lot of memory to avoid a loop where you just iteratively create coordindices.  

For the get_X_array(), did you consider using something simpler like the scipy sparse implementations, which are efficient and well tested?  

For example, if you create a [coo_matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html#scipy.sparse.coo_matrix) and then call [toarray()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spmatrix.toarray.html#scipy.sparse.spmatrix.toarray), you could dispense with this entire function and simplify get_X_array().

Reply:
I did think about using coo_matrix, but it doesn't quite do what we want.
We want to compress out rows and columns that were not selected.  The toarray function returns a dense array with a shape like (max row selected, max col selected).  coo_matrix is also not scriptable, so you can't slice into it either.  I could use coo_matrix, then convert to a 2d numpy array, then slice out of that.  The code would be simpler, but it would create a large dense matrix as an intermediate step.  Please let me know if you see a path to use coo matrix that doesn't blow up memory.

This approach does not use a lot of memory.   Assuming that we want to return a dense submatrix (what the frontend currently expects), the amount of temporary memory used here is less than the memory of the output array.  memory is O(nrows) or O(ncols).  The amount of compute is just O(nnz).   I did write this to avoid loop, and so maybe it's not so obvious from reading the code.   But the basic operation isn't too complicated.

Say there is a matrix like this:

```
0  1  2  0  3  0
4  0  5  0  6  0  
0  0  7  8  0  9
```

And we wanted to select all the rows,  and cols = 0, 2, 5
The output matrix should be:

```
0  2  0
4  5  0
0  7  9
```

From tiledb we get:
obs = [0, 1, 1, 2, 2]
var =  [2, 0, 2, 2, 5]
data = [2,4,5,7,9]

So the challenge is how to transform the three arrays (obs, var, data) to the 2d output matrix.

The first step is to create a mapping from X var coords to the output columns:
e.g. map 0->0, 2->1, 5->2.

```
# coord_mask = [1,0,1,0,0,1]
indices = np.where(coord_mask)[0]    # indices = [0,2,5]
ncoord = indices.shape[0]                   # ncoord = 3
maprange = np.arange(ncoord)          # maprange = [0,1,2]
mapindex = np.zeros(indices[-1] + 1, dtype=int)
mapindex[indices] = maprange          # mapindex = [0,0,1,0,0,2]
```

next map the var indices to the output columns:  [2,0,2,2,5] -> [1,0,1,1,2]
```
coordindices = mapindex[coord_data]. # coordindices = [1,0,1,1,2]
```

Now the obs and var coordinates are in right space, and can be used to broadcast to the 
dense output array:

```
densedata = np.zeros((nrows, ncols), dtype=self.get_X_array_dtype())
densedata[obsindices, varindices] = data[""]
```

Reply:
this is helpful.  I did eventually puzzle it out when doing the review, but given that you have done this super nice writeup, maybe add it to a comment in the code for the next person :-)

Reply:
Sounds good.  comment updated.